src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Net.Http.HPack;
+using System.Net.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+// Some stuff to do:
+
+// * Handle errors/shutdown, both graceful and abortive, at connection level and per-stream
+// * Handle request cancellation
+// * Handle request body
+// * Handle frame limit and sending CONTINUATION on request headers
+// * Flow control (both inbound and outbound)
+// * Various TODOs below
+
+namespace System.Net.Http
+{
+    // TODO: This should probably implement IDisposable.
+    internal sealed class Http2Connection : HttpConnectionBase
+    {
+        private readonly SslStream _stream;
+
+        private ArrayBuffer _incomingBuffer;
+        private ArrayBuffer _outgoingBuffer;
+
+        private readonly HPackDecoder _hpackDecoder;
+
+        private readonly ConcurrentDictionary<int, Http2Stream> _httpStreams;
+
+        private readonly SemaphoreSlim _writerLock;
+
+        private int _nextStream;
+        private bool _expectingSettingsAck;
+
+        private const int MaxStreamId = int.MaxValue;
+
+        private static readonly byte[] s_http2ConnectionPreface = Encoding.ASCII.GetBytes("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n");
+
+        private const int InitialBufferSize = 4096;
+
+        public Http2Connection(SslStream stream)
+        {
+            _stream = stream;
+            _incomingBuffer = new ArrayBuffer(InitialBufferSize);
+            _outgoingBuffer = new ArrayBuffer(InitialBufferSize);
+
+            _hpackDecoder = new HPackDecoder();
+
+            _httpStreams = new ConcurrentDictionary<int, Http2Stream>();
+
+            _writerLock = new SemaphoreSlim(1);
+
+            _nextStream = 1;
+        }
+
+        public async Task SetupAsync()
+        {
+            // Send connection preface
+            _outgoingBuffer.EnsureWriteSpace(s_http2ConnectionPreface.Length);
+            s_http2ConnectionPreface.AsSpan().CopyTo(_outgoingBuffer.WriteBytes);
+            _outgoingBuffer.Commit(s_http2ConnectionPreface.Length);
+
+            // Send empty settings frame
+            WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.None, 0));
+
+            // TODO: We should disable PUSH_PROMISE here.
```

Comment:
That's a requirement, right? i.e. if we don't do it, the server may push stuff to us, which we'll then fail on?

Reply:
Yes.